modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/db/Pool.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2012-2021 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.rdbloader.db
+
+import cats.implicits._
+
+import cats.effect.Concurrent
+import cats.effect.concurrent.{Ref, Semaphore}
+
+import fs2.concurrent.Queue
+
+/**
+ * A connection pool like entity, managing acquisition and use of several
+ * resources `R`. The `use` function is completely transparent to user
+ * code and identical to `Resource#use`, but in case of `Pool` a function
+ * would do one of the following:
+ * 1. Receive a free pre-allocated resource
+ * 2. Trigger creation of a new one if capacity allows
+ * 3. Block (semantically) until other fibers release a resource if capacity
+ * doesn't allow
+ *
+ * @tparam F an effect type, usually `IO`
+ * @tparam R a resource type, such as DB
+ */
+trait Pool[F[_], R] {
+  def use[O](f: R => F[O]): F[O]
+}
+
+object Pool {
+
+  def createQ[F[_]: Concurrent, R](acquire: F[R], release: R => F[Unit], max: Int): F[Pool[F, R]] = {
+    val resourceP = acquire.map(res => ResourceP(res, release(res)))
+    for {
+      resourceQueue <- Queue.bounded[F, ResourceP[F, R]](max)
```

Comment:
I guess this could safely be an unbounded queue?  Because it is the semaphore guarantees the number of resources, not the queue size.

I'm not saying to change it - just making sure I understand correctly.

Reply:
Yes, I also thought about it, but decided to keep as another layer of defense against exceeding `max`.

This is a philosophical question though whether I should be adding layers of uncertainty on top of something I don't 100% certain about.